### PR TITLE
Reduce prefix len for Evm + DbAccount storage

### PR DESCRIPTION
**Pull request by kpp**
# Description

- [x] Add `#[state(rename = "new_name")]` to shorten Evm states. Evm accounts -> a, code -> c, last_l1_hash -> l, latest_block_hashes -> h. DbAccount storage -> s, keys -> k.
So prefixes now are: `Evm/a`, `Evm/c`, `Evm/l`, `Evm/h`, `Evm/s` and `Evm/k`.
- [x] Add `#[memory]` to keep Evm block_env and pending_transactions out of storage writes.
- [x] Remove Account.info and Account.code when account is removed
- [x] Fix l1 diff size calc when account is removed
- [x] Fix l1 diff size calc for storage keys

## Linked Issues
- Fixes # (issue, if applicable)
- Related to # (issue) 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

### CODE REVIEW

File: crates/evm/src/evm/handler.rs [hunk 1]
```diff
@@ -23,18 +24,23 @@ use tracing::instrument;
 use crate::system_events::SYSTEM_SIGNER;
 use crate::{BASE_FEE_VAULT, L1_FEE_VAULT};
 
-const DB_ACCOUNT_SIZE: usize = 74;
+/// Eoa size is reduced because code_hash for eoas are None on state diff, converted to empty Keccak  internally for evm operations
+const DB_ACCOUNT_SIZE_EOA: usize = 42;
```
Comment (eyusufatik) [inline crates/evm/src/evm/handler.rs:28]: 74 - 32 + 1 (for none) = 43???

File: crates/evm/src/evm/handler.rs [hunk 2]
```diff
@@ -23,18 +24,23 @@ use tracing::instrument;
...
+const DB_ACCOUNT_SIZE_CONTRACT: usize = 75;
 
-/// Normally db account key is: 24 bytes of prefix + 1 byte for size of remaining data + 20 bytes of address = 45 bytes
+/// Normally db account key is: 6 bytes of prefix ("Evm/a/") + 1 byte for size of remaining data + 20 bytes of address = 27 bytes
 /// But we already add address size to diff size, so we don't need to add it here
```
Comment (eyusufatik) [inline crates/evm/src/evm/handler.rs:32]: should we also change this behaviour maybe?

Comment (eyusufatik) [reply crates/evm/src/evm/handler.rs:32]: just adding it up here, maybe code gets smaller

File: crates/evm/src/evm/handler.rs [hunk 3]
```diff
@@ -503,6 +509,15 @@ fn calc_diff_size<EXT, DB: Database>(
     }
 
     for (addr, account) in account_changes {
+        let db_account_size = {
+            let account = &state[addr];
```
Comment (eyusufatik) [inline crates/evm/src/evm/handler.rs:513]: what's the type of state? are we sure we can access all accounts like this?

File: crates/evm/src/evm/handler.rs [hunk 4]
```diff
@@ -531,7 +548,7 @@ fn calc_diff_size<EXT, DB: Database>(
             // DbAccount size is added because when any of those changes the db account is written to the state
```
Comment (kpp) [inline crates/evm/src/evm/handler.rs:548]: Fix comments

File: crates/evm/src/evm/handler.rs [hunk 5]
```diff
@@ -512,15 +527,17 @@ fn calc_diff_size<EXT, DB: Database>(
...
             // Retrieve code from DB and apply its size
             if let Some(info) = db.basic(*addr)? {
                 if let Some(code) = info.code {
                     diff_size += code.len();
+                    diff_size += CODE_KEY_SIZE;
                 } else {
                     let code = db.code_by_hash(info.code_hash)?;
                     diff_size += code.len();
+                    diff_size += CODE_KEY_SIZE;
                 }
```
Comment (eyusufatik) [inline crates/evm/src/evm/handler.rs:541]: can you also make this whole if stuff looked up from journal.state?

File: crates/evm/src/evm/handler.rs [hunk 6]
```diff
@@ -512,15 +527,17 @@ fn calc_diff_size<EXT, DB: Database>(
 
             // All the nonce, balance and code_hash fields are updated and written to the state with DbAccount
             diff_size +=
-                (DB_ACCOUNT_SIZE + DB_ACCOUNT_KEY_SIZE) * NONCE_DISCOUNTED_PERCENTAGE / 100;
+                (db_account_size + DB_ACCOUNT_KEY_SIZE) * NONCE_DISCOUNTED_PERCENTAGE / 100;
 
             // Retrieve code from DB and apply its size
             if let Some(info) = db.basic(*addr)? {
                 if let Some(code) = info.code {
                     diff_size += code.len();
```
Comment (eyusufatik) [inline crates/evm/src/evm/handler.rs:535]: if we are destroting account, we shouldn't be charging code.len();

we should be charging dbaccount.keys.len() * (dbaccount.keys key size + dbaccount.storage keysize + 2 (for writing null) ) +

File: crates/evm/src/evm/handler.rs [hunk 7]
```diff
@@ -512,15 +527,17 @@ fn calc_diff_size<EXT, DB: Database>(
...
+                    diff_size += CODE_KEY_SIZE;
                 } else {
                     let code = db.code_by_hash(info.code_hash)?;
                     diff_size += code.len();
+                    diff_size += CODE_KEY_SIZE;
```
Comment (eyusufatik) [inline crates/evm/src/evm/handler.rs:540]: should be code key size + 1 (for writing null)

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: crates/evm/src/evm/handler.rs (hunk 1 above)
Comment (ercecan) [reply crates/evm/src/evm/handler.rs:28]: 1 byte len descriptor for code hash

File: crates/evm/src/evm/handler.rs (hunk 3 above)
Comment (ercecan) [reply crates/evm/src/evm/handler.rs:513]: it is hashmap, if it is journaled, which is what we need, then yes

File: crates/evm/src/evm/handler.rs (hunk 1 above)
Comment (eyusufatik) [reply crates/evm/src/evm/handler.rs:28]: But option<T> 's size is 1 on none

File: crates/evm/src/evm/handler.rs (hunk 3 above)
Comment (eyusufatik) [reply crates/evm/src/evm/handler.rs:513]: i wish we could use btreemap for that haha

File: crates/evm/src/evm/handler.rs (hunk 1 above)
Comment (ercecan) [reply crates/evm/src/evm/handler.rs:28]: It should be 73 but there is 1 byte length descriptor for code hash 0x20 + code hash
0x20 is the 1 extra byte

File: crates/evm/src/tests/hooks_tests.rs [hunk 8]
```diff
@@ -56,7 +56,18 @@ fn begin_soft_confirmation_hook_creates_pending_block() {
...
+    let b = evm
+        .get_block_by_number(
+            Some(alloy_eips::BlockNumberOrTag::Earliest),
+            None,
+            &mut working_set,
+        )
+        .unwrap()
+        .unwrap();
+    println!("{:?}", b.header.hash.unwrap());
+    println!("{:?}", b.header.state_root);
```
Comment (eyusufatik) [inline crates/evm/src/tests/hooks_tests.rs:70]: delete

File: crates/evm/src/lib.rs [hunk 9]
```diff
@@ -85,29 +85,35 @@ pub struct Evm<C: sov_modules_api::Context> {
...
     /// Used by the EVM to calculate the `blockhash` opcode.
-    #[state]
+    #[state(rename = "h")]
     pub(crate) latest_block_hashes: sov_modules_api::StateMap<U256, B256, BcsCodec>,
 
+    /// Native pending transactions. Used to store transactions that are not yet included in the block.
+    /// We use this in the sequencer to see which txs did not fail
+    #[state]
+    pub(crate) native_pending_transactions:
+        sov_modules_api::AccessoryStateVec<PendingTransaction, BcsCodec>,
```
Comment (eyusufatik) [inline crates/evm/src/lib.rs:115]: we can make these native as well

File: crates/sovereign-sdk/module-system/sov-modules-macros/tests/rpc/expose_rpc_first_generic_not_context.stderr [hunk 10]
```diff
@@ -52,7 +52,7 @@ note: required because it appears within the type `Runtime<S, C>`
 note: required by a bound in `sov_modules_api::DispatchCall`
   --> $WORKSPACE/crates/sovereign-sdk/module-system/sov-modules-core/src/module/dispatch.rs
    |
-   | pub trait DispatchCall: Send + Sync {
+   | pub trait : Send + Sync {
```
Comment (eyusufatik) [inline crates/sovereign-sdk/module-system/sov-modules-macros/tests/rpc/expose_rpc_first_generic_not_context.stderr:55]: should be a DispatchCall here probably

**Review by eyusufatik — APPROVED** [review]

**Pull request merged**